torch/utils/data/dataset.py

```diff
@@ -253,6 +254,42 @@ def __len__(self):
         return total
 
 
+class ShuffleDataset(IterableDataset[T_co]):
+    r"""Dataset shuffled from the original dataset.
+
+    This class is useful to shuffle an existing instance of an IterableDataset.
+    The buffer with buffer_size is filled with the elemetns from the dataset first,
+    and then pops out a piece of data randomly for each iteration.
+    For the buffer size smaller than 1, the buffer size is set to 1 representing 
+    non-shuffled dataset. For perfect shuffling, the buffer size is required to be 
+    greater or equal to the size of dataset.
+
+    Arguments:
+        dataset (IterableDataset): The original IterableDataset.
+        buffer_size (int): The buffer size for shuffling.
+    """
+    dataset: IterableDataset[T_co]
+    buffer_size: int
+    seed: Optional[int]
+
+    def __init__(self, dataset: IterableDataset[T_co], buffer_size: int, seed: Optional[int] = None) -> None:
+        super(ShuffleDataset, self).__init__()
+        self.dataset = dataset
+        self.buffer_size = buffer_size if buffer_size > 1 else 1
+        self.seed = seed
+
+    def __iter__(self) -> Iterator[T_co]:
+        buf: List[T_co] = []
+        if self.seed:
+            seed(self.seed)
```

Comment:
Two thoughts on this:
- This makes every iteration over `ShuffledDataset` to reset to the same seed over and over again, while I think that most users would expect that the first instantiation of `__iter__` would set the seed. So my proposal would be to only call `random.seed` at the first invocation.
- Another point is that I think it's probably better to just import `random` and use `random.randint` and `random.seed`, as `seed` is a very common / used word and could be overwritten in this code.